src/v/raft/replicate_batcher.cc

```diff
@@ -28,41 +28,57 @@ replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
 replicate_stages replicate_batcher::replicate(
```

Comment:
> When the timeout error is triggered during the first phase the item is
not cached and the buffers are freed immediately. If an exception happen
during the second phase processing then the replicate request isn't
canceled but the result future is resolved by the timeout timer.
Replicate batcher memory units are released only after the replicate
call finishes even if future was already resolved. This way in case of
replication being slow and timing out we will still be propagating
backpressure without allocating memory for pending requests.

awesome. could you also comment on when a replication might timeout but still complete? is that possible?

Reply:
yes it is possible, it may happen only if the timeout is triggered during the second phase, will add a comment to commit message.